nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/AzureStorageUtils.java

```diff
@@ -85,6 +86,20 @@
             .sensitive(true)
             .build();
 
+    public static final PropertyDescriptor ENDPOINT_SUFFIX = new PropertyDescriptor.Builder()
+            .name("storage-endpoint-suffix")
+            .displayName("Common Storage Account Endpoint Suffix")
+            .description(
+                    "Storage accounts in public Azure always use a common FQDN suffix. " +
+                    "Override this endpoint suffix with a different suffix in certain circumsances (like Azure Stack or non-public Azure regions). " +
+                    "The preferred way is to configure them through a controller service specified in the Storage Credentials property. " +
+                    "The controller service can provide a common/shared configuration for multiple/all Azure processors. Furthermore, the credentials " +
+                    "can also be looked up dynamically with the 'Lookup' version of the service.")
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
```

Comment:
If I read everything correctly in [`AzureStorageUtils`](https://github.com/apache/nifi/blob/56fc89fcdea0e3146f0fe4a28904995fbad3646d/nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/AzureStorageUtils.java#L159-L169), _Storage Credentials_ service takes precedence over _Storage Account Name_ and _Storage Account Key_ or _SAS Token_. Those are validated as either-ors in the processor though so the behavior is predictable.

What happens though if we set _Endpoint Suffix_ on both the processor and the service? I think it's going to use what's on the service, which could be confusing.

And what if you use an expression on the service but it's blank? I don't think it'll default to what's set in the processor, since if the service is set it'll always use the service. So it's not a fallback either.

*If* this is correct I think there are a couple of options:

1. Validate that either _Storage Credentials_ service or _Endpoint Suffix_ are in use and not both.
2. Create some kind of fallback behavior in [`createStorageCredentialsDetails`](https://github.com/apache/nifi/blob/56fc89fcdea0e3146f0fe4a28904995fbad3646d/nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/AzureStorageUtils.java#L159) (use the endpoint suffix from the service if it's non-blank, otherwise fallback to the endpoint suffix from the processor context).
3. Make a note in the docs and do nothing.

(2) is very flexible but pretty complicated. (3) seems like it'll lead people to make mistakes.

Reply:
I took the option and set StandardValidators.NON_BLANK_VALIDATOR to EndpointSuffix property.

Reply:
Storage Credential service provides more than just the Endpoint Suffix, a user could want to supply the endpoint suffix in the processor and the account key in the Storage Credential service.

Could the validator change so that it checks that there is no overlap on a property level?

Reply:
I would suggest Option 1 as either, not both.

Reply:
Yes. Option 1 is implemented.